### ISSUE TITLE
Add "inferred suite" feature

### ISSUE THREAD

**Issue by vlsi** [cbeust/testng#2606]
Currently, the only ways to describe a test suite are:

a) Write suite.xml file
b) Use programmatic API to create XmlSuite and run it

Both approaches result in an awkward user experience:

1. IDE support is half-working. For instance, if my suite consists of multiple test classes, I can't open one of them and launch from IDE directly. I need to find the relevant suite file
2. Every time a new class is created, it needs to be added to a suite, otherwise the test won't be executed. That results in excessive overheads, and it is prone to human errors (people might fail to include a test in a suite and think "all tests pass")
3. Programmatically added tests are not runnable from IDE
4. Browsing test class code does not clarify what the test class is doing, as it might be a part of a weird XML file with its own parameters, listeners, etc, etc.  Just as a comparison, Java source files have `import ..` section at the beginning, and it helps developers to understand "what does the code use".
5. Listeners declared in test class annotation are not scoped to the given class, and I believe, no-one really expects that. Do you have a use-case, when a listener 

---

What do you think if TestNG could group tests into suites automatically?

In other words:
1. User or build tool asks TestNG to execute a "subset of tests classes"
2. TestNG analyzes the classes, and detects `@AutomaticSuite(name = "...")` (or `@ExtendsSuite(SomeSuite.class)`) annotation on top of test classes and so it could group the classes in suites.
3. Listeners should be specified in the class annotation, or in superclass/superinterface, or in composite annotations
4. Guice modules should be specified in the class annotation or inherited in a similar way via superclass/superinterface chain, etc

**Comment by juherr** [comment]
The suite is the root object of the TestNG model: you'll always run a suite, even if you run a test.

The TestNG API should be provided to allow you to convert whatever you want into a suite description.
Currently, you can run by CLI, code, or file (XML or YAML)

Why not add an extra SPI that will allow you to describe another format?

**Comment by vlsi** [comment]
Gradle uses `TestNG#setTestClasses(String[])`: https://github.com/cbeust/testng/blob/bb2bf12621866a7e03200eb235b3b62074bca7df/testng-core/src/main/java/org/testng/TestNG.java#L439-L447,  see Gradle https://github.com/gradle/gradle/blob/84f8ad994ec2808e053aa7d1a33d682237111974/subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/testng/TestNGTestClassProcessor.java#L135-L139

So the API to launch a given test class is already there.

**Comment by vlsi** [comment]
However, it is a valid question what should happen when IDE or build tool builds a suite programmatically (e.g. `new XmlSuite(..)`)

**Comment by vlsi** [comment]
Having thought a bit on this I incline that it might be easier to create `suite.xml` generator based on classfile analysis.

In other words, I could scan class files, group them by compatibility (e.g. I might have my own annotations describing if different tests can't be a part of a suite due to incompatible requirements), and then I could write one or several suite.xml files.
